### PR TITLE
chunk: append `offset` when elem len in dst column is variable in `appendCellByCell` function

### LINKED ISSUE #52159: probeWorker[1] meets error: runtime error: index out of range [431] with length 1
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
First execute the following valid.sql
[valid.txt](https://github.com/pingcap/tidb/files/14772480/valid.txt)

Then a crash occurs when executing the error.sql below
[error.txt](https://github.com/pingcap/tidb/files/14772482/error.txt)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
Expect no crashes

### 3. What did you see instead (Required)

```
err="probeWorker[1] meets error: runtime error: index out of range [431] with length 1
github.com/pingcap/tidb/pkg/executor.(*probeWorker).handleProbeWorkerPanic
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/join.go:418
github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1.1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/wait_group_wrapper.go:168
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:914
runtime.goPanicIndex
	/usr/local/go/src/runtime/panic.go:114
github.com/pingcap/tidb/pkg/util/chunk.appendCellByCell
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/chunk/chunk.go:421
github.com/pingcap/tidb/pkg/util/chunk.(*Chunk).AppendPartialRowByColIdxs
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/chunk/chunk.go:408
github.com/pingcap/tidb/pkg/util/chunk.(*Chunk).AppendRowByColIdxs
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/chunk/chunk.go:389
github.com/pingcap/tidb/pkg/executor.(*baseJoiner).makeJoinRowToChunk
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/joiner.go:239
github.com/pingcap/tidb/pkg/executor.(*innerJoiner).tryToMatchInners
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/joiner.go:976
github.com/pingcap/tidb/pkg/executor.(*probeWorker).joinMatchedProbeSideRow2Chunk
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/join.go:942
github.com/pingcap/tidb/pkg/executor.(*probeWorker).join2Chunk
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/join.go:1058
github.com/pingcap/tidb/pkg/executor.(*probeWorker).runJoinWorker
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/join.go:524
github.com/pingcap/tidb/pkg/executor.(*HashJoinExec).fetchAndProbeHashTable.func2
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/executor/join.go:401
github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithRecover.func1
	/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/wait_group_wrapper.go:172
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1650"

```
### 4. What is your TiDB version? (Required)

```
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                 |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.5.1
Edition: Community
Git Commit Hash: 7d16cc79e81bbf573124df3fd9351c26963f3e70
Git Branch: heads/refs/tags/v7.5.1
UTC Build Time: 2024-02-27 14:28:32
GoVersion: go1.21.6
Race Enabled: false
Check Table Before Drop: false
Store: tikv |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
<!-- Paste the output of SELECT tidb_version() -->

We are the BASS team from the School of Cyber Science and Technology at Beihang University. Our main focus is on system software security, operating systems, and program analysis research, as well as the development of automated program testing frameworks for detecting software defects. Using our self-developed database vulnerability testing tool, we have identified the above-mentioned vulnerabilities in TiDB that may lead to database crashes.

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #52159

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/util/chunk/chunk_test.go [hunk 1]
```diff
@@ -724,6 +724,24 @@ func TestToString(t *testing.T) {
...
+	srcCol.AppendInt64(3)
+
+	appendCellByCell(dstCol, srcCol, 0)
+	appendCellByCell(dstCol, srcCol, 1)
+	appendCellByCell(dstCol, srcCol, 2)
+
+	// Test is success when there is no panic
+	dstCol.GetString(0)
+	dstCol.GetString(1)
+	dstCol.GetString(2)
```
Comment (SeaRise) [inline pkg/util/chunk/chunk_test.go:742]: how about asserting data here, despite it is unrelated to this issue.

**Review by SeaRise — APPROVED** [review]
lgtm

File: pkg/util/chunk/chunk_test.go (hunk 1 above)
Comment (xzhangxian1008) [reply pkg/util/chunk/chunk_test.go:742]: > how about asserting data here, despite it is unrelated to this issue.

I think it's needless as we only need to ensure that there is no panic.

**Comment by xzhangxian1008** [comment]
Close it. Not a executor bug but optimizer.

**Pull request closed without merging**